packages/builders/src/messages/AllowedMentions.ts

```diff
@@ -0,0 +1,172 @@
+import type { JSONEncodable } from '@discordjs/util';
+import type { AllowedMentionsTypes, APIAllowedMentions, Snowflake } from 'discord-api-types/v10';
+import { normalizeArray, type RestOrArray } from '../util/normalizeArray.js';
+import { validate } from '../util/validation.js';
+import { allowedMentionPredicate } from './Assertions.js';
+
+/**
+ * A builder that creates API-compatible JSON data for allowed mentions.
+ */
+export class AllowedMentionsBuilder implements JSONEncodable<APIAllowedMentions> {
+	private readonly data: Partial<APIAllowedMentions>;
+
+	/**
+	 * Creates new allowed mention builder from API data.
+	 *
+	 * @param data - The API data to create this attachment with
+	 */
+	public constructor(data: Partial<APIAllowedMentions> = {}) {
+		this.data = structuredClone(data);
+	}
+
+	/**
+	 * Sets the types of mentions to parse from the content.
+	 *
+	 * @param parse - The types of mentions to parse from the content
+	 */
+	public setParse(...parse: RestOrArray<AllowedMentionsTypes>): this {
+		this.data.parse = normalizeArray(parse);
+		return this;
+	}
+
+	/**
+	 * Clear the types of mentions to parse from the content.
+	 */
+	public clearParse(): this {
+		this.data.parse = [];
+		return this;
+	}
+
+	/**
+	 * Sets the roles to mention.
+	 *
+	 * @param roles - The roles to mention
+	 */
+	public setRoles(...roles: RestOrArray<Snowflake>): this {
+		this.data.roles = normalizeArray(roles);
+		return this;
+	}
+
+	/**
+	 * Adds roles to mention.
+	 *
+	 * @param roles - The roles to mention
+	 */
+	public addRoles(...roles: RestOrArray<Snowflake>): this {
+		this.data.roles ??= [];
+		this.data.roles.push(...normalizeArray(roles));
+
+		return this;
+	}
+
+	/**
+	 * Removes, replaces, or inserts roles.
+	 *
+	 * @remarks
+	 * This method behaves similarly
+	 * to {@link https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/Array/splice | Array.prototype.splice()}.
+	 *
+	 * It's useful for modifying and adjusting order of the already-existing roles.
+	 * @example
+	 * Remove the first role:
+	 * ```ts
+	 * allowedMentions.spliceRoles(0, 1);
+	 * ```
+	 * @example
+	 * Remove the first n role:
+	 * ```ts
+	 * const n = 4;
+	 * allowedMentions.spliceRoles(0, n);
+	 * ```
+	 * @example
+	 * Remove the last role:
+	 * ```ts
+	 * allowedMentions.spliceRoles(-1, 1);
+	 * ```
+	 * @param index - The index to start at
+	 * @param deleteCount - The number of roles to remove
+	 * @param roles - The replacing role ids
+	 */
+	public spliceRoles(index: number, deleteCount: number, ...roles: RestOrArray<Snowflake>): this {
```

Comment:
Let's add in clearRoles too?

Reply:
We don't do this generally because you can just `setRoles()`, this has been the convention since I did builders v2 and other classes follow the pattern (embeds, etc)

Reply:
Then maybe clearParse should be dipped?